cmudb/extensions/tscout/main.c

```diff
@@ -0,0 +1,224 @@
+#include <time.h>
+
+// clang-format off
+// Extension magic.
+#include "postgres.h"
+#include "fmgr.h"
+// clang-format on
+
+#include "commands/createas.h"
+#include "commands/explain.h"
+#include "optimizer/planner.h"
+#include "parser/parsetree.h"
+#include "utils/builtins.h"
+
+#include "../../../src/include/tscout/marker.h"
+#include "operating_unit_features.h"
+
+PG_MODULE_MAGIC;
+void _PG_init(void);
+void _PG_fini(void);
+
+static void ExplainOneQueryWrapper(Query *query, int cursorOptions, IntoClause *into, ExplainState *es,
+                                   const char *queryString, ParamListInfo params, QueryEnvironment *queryEnv);
+static void WalkPlan(Plan *plan, ExplainState *es);
+static void explainXs(Plan *node, ExplainState *es);
+static size_t get_field_size(c_type type);
+
+static ExplainOneQuery_hook_type chain_ExplainOneQuery = NULL;
+
+void _PG_init(void) {
+  elog(LOG, "Initializing extension");
+
+  // Chain the Extension Explain Query Wrapper to the head of the chain.
+  chain_ExplainOneQuery = ExplainOneQuery_hook;
+  ExplainOneQuery_hook = ExplainOneQueryWrapper;
+
+  // Init logic (like setting flags) go here
+}
+
+void _PG_fini(void) {
+  // Clean up logic goes here
+  ExplainOneQuery_hook = chain_ExplainOneQuery;
+  elog(DEBUG1, "Finishing extension");
+}
+
+static void ExplainOneQueryWrapper(Query *query, int cursorOptions, IntoClause *into, ExplainState *es,
+                                   const char *queryString, ParamListInfo params, QueryEnvironment *queryEnv) {
+  PlannedStmt *plan;
+  QueryDesc *queryDesc;
+  instr_time plan_start, plan_duration;
+  int eflags = 0;
+
+  if (chain_ExplainOneQuery) {
+    chain_ExplainOneQuery(query, cursorOptions, into, es, queryString, params, queryEnv);
+  }
+
+  // Postgres does not expose an interface to call into the standard ExplainOneQuery.
+  // Hence, we duplicate the operations performed by the standard ExplainOneQuery ie
+  // calling into the standard planner.
+  // A non-standard planner can be hooked in, in the the future.
+  INSTR_TIME_SET_CURRENT(plan_start);
+  plan = (planner_hook ? planner_hook(query, queryString, cursorOptions, params)
+                       : standard_planner(query, queryString, cursorOptions, params));
+
+  INSTR_TIME_SET_CURRENT(plan_duration);
+  INSTR_TIME_SUBTRACT(plan_duration, plan_start);
+
+  TS_MARKER("TS_MARKER <NAME>");
+
+  if (es->format == EXPLAIN_FORMAT_TSCOUT) {
+    queryDesc =
+        CreateQueryDesc(plan, queryString, InvalidSnapshot, InvalidSnapshot, None_Receiver, params, queryEnv, 0);
+
+    if (es->analyze)
+      eflags = 0;
+    else
+      eflags = EXEC_FLAG_EXPLAIN_ONLY;
+    if (into) eflags |= GetIntoRelEFlags(into);
+
+    // Run the executor
+    ExecutorStart(queryDesc, eflags);
+    // This calls into initPlan() which populates the plan tree.
+    // TODO: Create a hook to executor start.
+
+    // Finally, walk through the plan, dumping its out in a separate top-level group.
```

Comment:
Style nit: I find it helpful to try to avoid using words like `it` in my comments. Instead, prefer the name of the relevant object, if any. This removes ambiguity as to what "it" refers to.

```suggestion
    // Finally, walk through the plan, dumping the output of this extension into a separate top-level group.
```